R/derive_var_trtemfl.R

```diff
@@ -186,56 +249,133 @@ derive_var_trtemfl <- function(dataset,
       end_date,
       trt_start_date,
       trt_end_date,
       initial_intensity,
       intensity
     )
-  )
+  if (!is.null(group_var)) {
+    required_vars <- c(required_vars, enexpr(group_var))
+  }
+  assert_data_frame(dataset, required_vars = required_vars)
+
+  # Assert date variables
   assert_date_var(dataset, var = !!start_date)
   assert_date_var(dataset, var = !!end_date)
   assert_date_var(dataset, var = !!trt_start_date)
   if (!is.null(trt_end_date)) {
     assert_date_var(dataset, var = !!trt_end_date)
   }
 
+  # Compute TEAE flag based on the presence of group_var
+
   if (is.null(end_window)) {
     end_cond <- expr(TRUE)
   } else {
     if (is.null(trt_end_date)) {
-      abort(paste(
-        "`end_window` argument was specified but not `trt_end_date`",
-        "Either both or none of them must be specified.",
-        sep = "\n"
-      ))
+      abort(
+        paste(
+          "`end_window` argument was specified but not `trt_end_date`",
+          "Either both or none of them must be specified.",
+          sep = "\n"
+        )
+      )
     }
     if (ignore_time_for_trt_end) {
-      end_cond <- expr(is.na(!!trt_end_date) |
-        date(!!start_date) <= date(!!trt_end_date) + days(end_window))
+      end_cond <- expr(
+        is.na(!!trt_end_date) |
+          date(!!start_date) <= date(!!trt_end_date) + days(end_window)
+      )
     } else {
-      end_cond <- expr(is.na(!!trt_end_date) | !!start_date <= !!trt_end_date + days(end_window))
+      end_cond <-
+        expr(is.na(!!trt_end_date) |
+          !!start_date <= !!trt_end_date + days(end_window))
     }
   }
 
-  if (is.null(intensity)) {
-    worsening_cond <- expr(FALSE)
+  if (is.null(group_var)) {
+    if (is.null(intensity)) {
+      worsening_cond <- expr(FALSE)
+    } else {
+      worsening_cond <-
+        expr(
+          !!start_date < !!trt_start_date &
+            (
+              !!initial_intensity < !!intensity |
+                is.na(!!initial_intensity) | is.na(!!intensity)
+            )
+        )
+    }
   } else {
-    worsening_cond <-
-      expr(!!start_date < !!trt_start_date &
-        (!!initial_intensity < !!intensity | is.na(!!initial_intensity) | is.na(!!intensity)))
+    # Determine if there are multiple records per AE event
+    multiple_records <- dataset %>%
+      group_by(USUBJID, !!group_var) %>%
+      summarise(num_records = n(), .groups = "drop") %>%
+      filter(num_records > 1) %>%
+      nrow() > 0
+
+    # Compute TEAE flag based on the presence of multiple records
+    if (multiple_records) {
+      # Step 1: Identify worsened records after TRTSDT
+      # Step 2: Get the first worsen date after TRTSDT
+      dataset <- dataset %>%
+        arrange(USUBJID, !!group_var, !!start_date) %>%
+        group_by(USUBJID, !!group_var) %>%
+        mutate(worsen_date = case_when(
+          start_date >= trt_start_date &
+            (!!intensity > !is.na(!!initial_intensity)) ~ as_datetime(!!start_date),
```

Comment:
lag() gives NA for the first record, so it would miss those where 1st record is a worsened one (tried a random example).
![image](https://github.com/pharmaverse/admiral/assets/134052893/c97da955-8b75-4447-98e5-483d5a482eca)

I was considering that AEITOXGR will remain consistent for a group ID, it might be more convenient to compare with AEITOXGR to identify the worsened records. What do you think?

Reply:
I think we need to clarify the definition of "worsened" - do you think the 3rd and 4th records should be flagged as TEAE in `comp` from [here](https://github.com/pharmaverse/admiral/pull/2362#discussion_r1534847952)?


1. The change of `AETOXGR` for this episode of AE: 3->1->2->1 
2. The 3rd record has `AETOXGR=2`, `AEITOXGR=3`, and its previous record has `AETOXGR=1`.  If we compare it with the previous record, the 3rd record is worsened, but it's improving if we compare it with `AEITOXGR`.
3. If we assume `AEITOXGR` is derived from the first record within each `group_var`, i.e. the first record has `AEITOXGR = AETOXGR`, then it's okay to have `NA` for the first record as no comparison is necessary. 
 

The `TEAE_expected` values below are just my opinion, please let me know your thoughts. Thanks for delving into this!
```r
# A tibble: 7 × 10
  USUBJID ASTDTM     AENDTM     AEITOXGR AETOXGR AEGRPID TRTSDTM    TRTEDTM    TEAE_expected TEAE_derived
  <chr>   <date>     <date>     <chr>    <chr>   <chr>   <date>     <date>     <chr>         <chr>       
1 1       2021-12-30 2022-01-14 3        3       1       2022-01-01 2022-04-30 NA            NA          
2 1       2022-01-05 2022-06-01 3        1       1       2022-01-01 2022-04-30 NA            Y           
3 1       2022-01-10 2022-01-11 3        2       1       2022-01-01 2022-04-30 Y             Y           
4 1       2022-01-13 2022-03-01 3        1       1       2022-01-01 2022-04-30 Y             Y             
```

Reply:
Thanks @kaz462 for clarifying my doubts and apologies for the delayed response -

1. I agree with the 3rd record being flagged as TEAE since its AE severity worsened from 1>2 
2. For 4th record, I think it shouldn't be flagged since the severity improved within the trt window (I remember we agreed before, that any record after the worsened record should be flagged as TEAE, but I was referring to this paper again and it mentions how to handle this scenario. Pls refer to scenario 4 from this doc and let me know if you agree to this- [TEAE: Did I flag it right?](https://www.pharmasug.org/proceedings/2015/PO/PharmaSUG-2015-PO01.pdf)

Reply:
>  Pls refer to scenario 4 and let me know if you agree to this- [TEAE: Did I flag it right?](https://www.pharmasug.org/proceedings/2015/PO/PharmaSUG-2015-PO01.pdf)

I agree with not flagging the 2nd record (improved). For the 4th record, this type (worsened then improved) is flagged as TEAE in the study I'm working on, similar to the `ASEQ =6` record [discussed here](https://github.com/pharmaverse/admiral/issues/2302#issuecomment-1936312763). We may need to re-visit this decision: 

- Once a record within a grouped AE is flagged as a TEAE, then any subsequent record during treatment should also be a TEAE regardless of the severity

Options to consider are: 
1. stay with the current decision
2. the 4th record shouldn't be flagged as it has improved compared to the previous one
3. add more flexibility to accommodate the above two scenarios
     - details to be discussed...   

```r
# A tibble: 7 × 10
  USUBJID ASTDTM     AENDTM     AEITOXGR AETOXGR AEGRPID TRTSDTM    TRTEDTM 
  <chr>   <date>     <date>     <chr>    <chr>   <chr>   <date>     <date>     
1 1       2021-12-30 2022-01-14 3        3       1       2022-01-01 2022-04-30   
2 1       2022-01-05 2022-06-01 3        1       1       2022-01-01 2022-04-30  
3 1       2022-01-10 2022-01-11 3        2       1       2022-01-01 2022-04-30       
4 1       2022-01-13 2022-03-01 3        1       1       2022-01-01 2022-04-30          
```
@pharmaverse/admiral I added this topic to the [next meeting agenda](https://github.com/pharmaverse/admiral/discussions/2169#discussioncomment-8929601)

Reply:
In the Roche standard an AE is considered treatment emergent if the grade at any time at or after treatment start is greater than the grade at the last observation before treatment start. However, the granular grading information is collected in `FA`. In `AE` there is only a single observation per AE.

Reply:
@bundfussr Thanks! How would you flag the 3rd and 4th records above?
- the grade of the 3rd record (`AETOXGR=2`) is less than the grade at the last observation before treatment start (`AETOXGR=3`), but greater than its previous record (`AETOXGR=1`)  - it's flagged at Bayer

> In `AE` there is only a single observation per AE.

If Roche doesn't need the `group_var` feature and there are no general rules for the 3rd and 4th records above, should we implement it at the company level instead?

Reply:
@ashachakma can you look into what GSK's standards are around this issue please?  

@kaz462 if we can build this function argument as option (my understanding) then I think we can keep in admiral-core.  I don't think Bayer is the only company doing this??  @pharmaverse/admiral_comm any thoughts

Reply:
> @ashachakma can you look into what GSK's standards are around this issue please?

In GSK, AE data are collected as a single AE episode, so no standards are present for this scenario.

Reply:
> How would you flag the 3rd and 4th records above?

At Roche the AEs are not flagged at that granular level. In general, I think it depends on the analysis. However, I am not aware of any analysis where the 4th record would be used. I would use the worst grade only.

Reply:
Did you consider the scenario where the grades are 3, 1, 2 and only the third one is after treatment start? At Roche we would consider the AE as treatment emergent but I guess the current implementation wouldn't flag any of the records.

Reply:
> Did you consider the scenario where the grades are 3, 1, 2 and only the third one is after treatment start?

Yes. Only the 3rd one (grade=2) will be flagged based on [this algorithm](https://github.com/pharmaverse/admiral/issues/2302#issuecomment-1936312763), as it worsened compared to the last observation before treatment start (grade=1).

Reply:
@kaz462 , yes, you are right. However, I wonder if the algorithm is correct. Consider for example:
```
> adae2 <- expected2 <- tribble(
+     ~USUBJID, ~ASTDTM, ~AENDTM, ~AEITOXGR, ~AETOXGR, ~AEGRPID,
+     # starting before treatment and ending during treatment
+     "1", "2021-12-30T20:00", "2022-01-04T11:00", "2", "2", "2",
+     "1", "2022-01-05T20:15", "2022-01-09T01:23", "2", "3", "2",
+     "1", "2022-01-10T20:15", "2022-03-01T01:23", "2", "1", "2",
+ ) %>%
+     mutate(
+         ASTDTM = ymd_hm(ASTDTM),
+         AENDTM = ymd_hm(AENDTM),
+         TRTSDTM = ymd_hm("2022-01-10T01:01"),
+         TRTEDTM = if_else(USUBJID == "1", ymd_hm("2022-04-30T23:59"), ymd_hms(""))
+     )
> derive_var_trtemfl(
+     adae2,
+     new_var = TRTEMFL,
+     trt_end_date = TRTEDTM,
+     end_window = 10,
+     initial_intensity = AEITOXGR,
+     intensity = AETOXGR,
+     group_var = AEGRPID
+ ) %>% select(ASTDTM, AENDTM, AEITOXGR, AETOXGR, AEGRPID, TRTEMFL, TRTSDTM)
# A tibble: 3 × 7
  ASTDTM              AENDTM              AEITOXGR AETOXGR AEGRPID TRTEMFL TRTSDTM            
  <dttm>              <dttm>              <chr>    <chr>   <chr>   <chr>   <dttm>             
1 2021-12-30 20:00:00 2022-01-04 11:00:00 2        2       2       NA      2022-01-10 01:01:00
2 2022-01-05 20:15:00 2022-01-09 01:23:00 2        3       2       NA      2022-01-10 01:01:00
3 2022-01-10 20:15:00 2022-03-01 01:23:00 2        1       2       Y       2022-01-10 01:01:00
```
I think the third observation shouldn't be flagged because the worsening happened before treatment start.

What do you think?

Reply:
> I think the third observation shouldn't be flagged because the worsening happened before treatment start.

Agree the 3rd record shouldn't be flagged, as it's an improved one compared to its previous record, it aligns with [this algorithm](https://github.com/pharmaverse/admiral/issues/2302#issuecomment-1936312763) (the current PR has some deviations) 
Btw, shoud `AEITOXGR` be 3 in the example above? so the 3rd record is comparing with `AETOXGR = 3` instead of `AETOXGR = 2`, no impact on this example though.

Reply:
If `AEITOXGR` is the initial toxicity, I would expect `AEITOXGR = 2`.

However, I wonder if we need to consider `AEITOXGR` at all if `group_var` is specified. Couldn't we just identify the first worsening after treatment start and flag all records at and after that?

Reply:
> If `AEITOXGR` is the initial toxicity, I would expect `AEITOXGR = 2`.

My understanding of `AEITOXGR` is derived from the last AE record before `TRTSDT`, as a pre-treatment status, but I'm not sure if there are cases that use the 1st record's `AETOXGR` before `TRTSDT` as `AEITOXGR`.

> However, I wonder if we need to consider `AEITOXGR` at all if `group_var` is specified. Couldn't we just identify the first worsening after treatment start and flag all records at and after that?

To determine if the first record has improved or worsened, a comparison between the first record's `AETOXGR` (after treatment start) and `AEITOXGR` is needed, what do you think? 
If we have a fixed algorithm to derive `AEITOXGR`, then we can derive it inside of the function, and users do not need to specifying `initial_intensity`.

Reply:
At Roche `AEITOXGR` is the initial grade, i.e., the grade observed when the AE started regardless if it was before or after treatment start. However, this is Roche specific, `AEITOXGR` is not a CDSIC SDTM variable.

Why do we need `AEITOXGR` if `group_var` is specified? We could select the last record before treatment start and all records thereafter. If any worsening (compared to the previous record) occurs, the record is flagged (and all records after that). Did I miss something?

Reply:
If we combine the following AEs into a single record -
```r
  USUBJID AEGRPID ASTDT      AENDT      AETOXGR TRTSDT     TRTEDT    
  <chr>   <chr>   <date>     <date>     <chr>   <date>     <date>    
1 1       2       2021-12-30 2022-01-04 2       2022-01-10 2022-04-30
2 1       2       2022-01-05 2022-01-09 3       2022-01-10 2022-04-30
3 1       2       2022-01-10 2022-03-01 3       2022-01-10 2022-04-30
```

1. Use the first record before treatment `AEITOXGR=2` as the initial intensity , the combined AE below should be flagged as TEAE, same for the 3rd AE above. 
```r
  USUBJID AEGRPID ASTDT      AENDT      AEITOXGR AETOXGR TRTSDT     TRTEDT     TRTEMFL
  <chr>   <chr>   <date>     <date>     <chr>    <chr>   <date>     <date>     <chr>  
1 1       2       2021-12-30 2022-03-01 2        3       2022-01-10 2022-04-30 Y  
```


2. Use the last record before treatment `AEITOXGR=3` as the initial intensity, the combined AE below shouldn't be flagged as TEAE, same for the 3rd AE above.
```r
  USUBJID AEGRPID ASTDT      AENDT      AEITOXGR AETOXGR TRTSDT     TRTEDT     TRTEMFL
  <chr>   <chr>   <date>     <date>     <chr>    <chr>   <date>     <date>     <chr>  
1 1       2       2021-12-30 2022-03-01 3        3       2022-01-10 2022-04-30 NA   
```

The current [derive_var_trtemfl](https://pharmaverse.github.io/admiral/reference/derive_var_trtemfl.html) determines worsened record by comparing the `intensity` with `initial_intensity`. Should `initial_intensity`/`AEITOXGR` be interpreted as pre-treatment status in the function, and with the same definition regardless of AE data collection method? 
When determining pre-treatment status, should we always use the last record before treatment, or allow flexibility?

Reply:
> Should `initial_intensity`/`AEITOXGR` be interpreted as pre-treatment status in the function, and with the same definition regardless of AE data collection method?

I would interpret `initial_intensity`/`AEITOXGR` as the grade/intensity when the AE started. I would use it only if a single record per AE is collected. If a separate records is collected for each change in grade/intensity, I would ignore `initial_intensity`/`AEITOXGR` and just use `intensity`/`AETOXGR`. This seems easier and clearer to me.

> When determining pre-treatment status, should we always use the last record before treatment, or allow flexibility?

Using the last record seems reasonable to me. I am not aware of a use case where I would deviate from this. Thus I would not allow flexibility for now.

Reply:
Determining TEAEs is definitely not as easy as it looks upon first glance.  I don't know if it will help, but the PHUSE/CS Working Group on TEAEs published this poster last fall- [](https://www.lexjansen.com/css-us/2023/POS_PP09.pdf).  It contains results of a survey they conducted, with responses from people in a variety of the disciplines involved in clinical trials, and as you can see, there were several scenarios without any sort of general agreement as to whether the AEs in question should be considered as treatment-emergent.  Maybe the best thing to do is make sure the situations where there was broad agreement are being handled correctly, and then pick a direction for the rest, but be sure it's documented somewhere how those have been resolved, and allow a way to override the default behavior since there isn't a standard industry approach in those cases.

Reply:
As discussed in the meeting, we have decided to implement the `group_var` feature based on [option 1](https://github.com/pharmaverse/admiral/pull/2362#discussion_r1543839585) for now.

We can cross-check the functionality with two white papers when they become available, and add features as needed later. @nbrucken17 please keep us posted, thanks :) 

https://advance.phuse.global/display/WEL/Treatment+Emergent+Definitions+Recommendations
https://advance.phuse.global/display/WEL/Adverse+Event+Collection+Recommendations